File: benchmark/lua/heap_sort.lua
local start = os.clock()
local arr = {}
for i = 1000, 1, -1 do table.insert(arr, i) end
local function heapify(n, idx)
    local largest = idx
    local left = 2 * idx + 1
    local right = 2 * idx + 2
    if left < n and arr[left + 1] > arr[largest + 1] then largest = left end
    if right < n and arr[right + 1] > arr[largest + 1] then largest = right end
    if largest ~= idx then
        arr[idx + 1], arr[largest + 1] = arr[largest + 1], arr[idx + 1]
        heapify(n, largest)
    end
end
local n = #arr
for j = math.floor(n / 2) - 1, 0, -1 do heapify(n, j) end
for k = n - 1, 1, -1 do
    arr[1], arr[k + 1] = arr[k + 1], arr[1]
    heapify(k, 0)
end
local finish = os.clock()
print("First: " .. arr[1])
print("Last: " .. arr[#arr])
print(string.format("Time: %.0fms", (finish - start) * 1000))

File: benchmark/python/heap_sort.py
import time
arr = list(range(1000, 0, -1))
start = time.time() * 1000
def heapify(n, idx):
    largest = idx
    left = 2 * idx + 1
    right = 2 * idx + 2
    if left < n and arr[left] > arr[largest]: largest = left
    if right < n and arr[right] > arr[largest]: largest = right
    if largest != idx:
        arr[idx], arr[largest] = arr[largest], arr[idx]
        heapify(n, largest)
n = len(arr)
for j in range(n // 2 - 1, -1, -1): heapify(n, j)
for k in range(n - 1, 0, -1):
    arr[0], arr[k] = arr[k], arr[0]
    heapify(k, 0)
end = time.time() * 1000
print(f"First: {arr[0]}")
print(f"Last: {arr[-1]}")
print(f"Time: {end - start:.0f}ms")

